src/virtual-machine.js

```diff
@@ -541,19 +543,77 @@ class VirtualMachine extends EventEmitter {
-                this.runtime.storage.get(id).dataFormat === 'svg') {
+        if (!id || !this.runtime || !this.runtime.storage) return null;
+        const format = this.runtime.storage.get(id).dataFormat;
+        if (format === this.runtime.storage.DataFormat.SVG) {
             return this.runtime.storage.get(id).decodeText();
+        } else if (format === this.runtime.storage.DataFormat.PNG ||
+                format === this.runtime.storage.DataFormat.JPG) {
+            return this.runtime.storage.get(id).encodeDataURI();
         }
+        log.error(`Unhandled format: ${this.runtime.storage.get(id).dataFormat}`);
         return null;
     }
 
+    /**
+     * Update a costume with the given bitmap
+     * @param {!int} costumeIndex - the index of the costume to be updated.
+     * @param {!ImageData} bitmap - new bitmap for the renderer.
+     * @param {!number} rotationCenterX x of point about which the costume rotates, relative to its upper left corner
+     * @param {!number} rotationCenterY y of point about which the costume rotates, relative to its upper left corner
+     * @param {!number} bitmapResolution 1 for bitmaps that have 1 pixel per unit of stage,
+     *     2 for double-resolution bitmaps
+     */
+    updateBitmap (costumeIndex, bitmap, rotationCenterX, rotationCenterY, bitmapResolution) {
+        const costume = this.editingTarget.getCostumes()[costumeIndex];
+        if (!(costume && this.runtime && this.runtime.renderer)) return;
+
+        costume.rotationCenterX = rotationCenterX;
+        costume.rotationCenterY = rotationCenterY;
+
+        // @todo: updateBitmapSkin does not take ImageData
+        const canvas = document.createElement('canvas');
+        canvas.width = bitmap.width;
+        canvas.height = bitmap.height;
+        const context = canvas.getContext('2d');
+        context.putImageData(bitmap, 0, 0);
+        
+        // Divide by resolution because the renderer's definition of the rotation center
+        // is the rotation center divided by the bitmap resolution
+        this.runtime.renderer.updateBitmapSkin(
+            costume.skinId,
+            canvas,
+            bitmapResolution,
+            [rotationCenterX / bitmapResolution, rotationCenterY / bitmapResolution]
+        );
+
+        // @todo there should be a better way to get from ImageData to a decodable storage format
+        canvas.toBlob(blob => {
+            const reader = new FileReader();
+            reader.addEventListener('loadend', () => {
+                const storage = this.runtime.storage;
+                costume.assetId = storage.builtinHelper.cache(
+                    storage.AssetType.ImageBitmap,
+                    storage.DataFormat.PNG,
+                    Buffer.from(reader.result)
+                );
+                costume.dataFormat = storage.DataFormat.PNG;
+                costume.bitmapResolution = bitmapResolution;
+                costume.size = [bitmap.width, bitmap.height];
```

Comment:
This `size` is used by the GUI to display the width/height of the costume. Do you think it should be scaled here by the bitmap resolution, or should the GUI scale it by bitmap resolution when displaying that info?

Reply:
I'm fine with having the GUI deal with it, just wondering if you had an opinion. This behavior _seems_ wrong (watch the size in the costume tile)

![size-switching](https://user-images.githubusercontent.com/654102/39491356-20f523bc-4d5a-11e8-916b-fbfd6419a82a.gif)

Reply:
Ohh yes you're right, I should fix that

So this is something @kchadha and I have been talking about a lot recently. Here is what we found: https://docs.google.com/document/d/1NFWOLRCgki9JKZkRVO8XUhyFR783fW0UTevUh8Usvts/edit?usp=sharing

VM, GUI, and paint generally expect to be working with true sizes and rotation centers. Should we change it so that the GUI expects to be getting its information manipulated by the VM the same way we do for the renderer? GUI and VM share the costume object, so it would be messy to try to switch. I think GUI should probably halve the size it displays in the container, so that the costume js object is internally consistent.

Reply:
Eeeeek. 
If I'm understanding the discussion above correctly, I say keep VM behavior the way it is right now, and have GUI do the math explicitly for display purposes.